### PR TITLE
feat: Create and restore backup (AR-1421)

### PR DESCRIPTION
**Pull request by trOnk12**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

- We want to allow the user to create and restore the backup.

### Solutions

1. Create a backup database file. 
2. Encrypt the backup  database file. ✅ 
3. Perform the restoring of the backup file :
https://github.com/wireapp/kalium/pull/1096 ✅ 

### Dependencies (Optional)
We had to pull in the libsodium KMP bindings library, in an effort to be able to use the `xChaCha20-Poly1305` stream encryption algorithm.

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

### Notes (Optional)
Another PR in AR will follow to this one.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: cryptography/src/commonMain/kotlin/com/wire/kalium/cryptography/backup/BackupCoder.kt [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            decodedHeader.format != format -> {
+                kaliumLogger.e("The backup format found in the backup file header is not a valid one")
+                INVALID_FORMAT
+            }
+            decodedHeader.version.toInt() < version.toInt() -> {
+                kaliumLogger.e("The backup version found in the backup file header is not a valid one")
+                INVALID_VERSION
+            }
+            else -> null
+        }
```
Comment (trOnk12) [inline cryptography/src/commonMain/kotlin/com/wire/kalium/cryptography/backup/BackupCoder.kt:47]: this is a nice candidate for a seperated function

File: cryptography/src/commonMain/kotlin/com/wire/kalium/cryptography/utils/ChaCha20Utils.kt [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                    secretStreamState.state,
+                    encryptedData,
+                    additionalInformation
+                ).let { it.decryptedData.toByteArray() to it.tag.toInt() }
+
+                decryptionBufferedSink.write(decryptedData)
+                decryptedDataSize += decryptedData.size
+
+                // Stop reading the file if we reach the end of the stream
+                if (tag == crypto_secretstream_xchacha20poly1305_TAG_FINAL) {
```
Comment (trOnk12) [inline cryptography/src/commonMain/kotlin/com/wire/kalium/cryptography/utils/ChaCha20Utils.kt:132]: can be moved to a local variable

File: cryptography/src/commonMain/kotlin/com/wire/kalium/cryptography/utils/ChaCha20Utils.kt [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        try {
+            val additionalInformation: UByteArray = BackupCoder.version.encodeToByteArray().toUByteArray()
+            val backupCoder = BackupCoder(userId, passphrase)
+            val (decodingError, header) = backupCoder.decodeHeader(encryptedDataSource)
+
+            // If there was an error decoding the header, we return it
+            decodingError?.let { return it to 0L }
+
+            // We need to read the ChaCha20 generated header prior to the encrypted backup file data to run some sanity checks
+            val chaChaHeaderKey = backupCoder.generateChaCha20Key(header).toUByteArray()
```
Comment (trOnk12) [inline cryptography/src/commonMain/kotlin/com/wire/kalium/cryptography/utils/ChaCha20Utils.kt:107]: it already returns UByteArray no ?

**Review by ohassine — CHANGES_REQUESTED** [review]
Great effort 💪 I  just left some comments about utils and some missing test

File: logic/src/commonJvmAndroid/kotlin/com.wire.kalium.logic/util/BackupUtils.kt [hunk 4]
```diff
@@ -0,0 +1,85 @@
+package com.wire.kalium.logic.util
```
[the remark this answers is not in the document]
Comment (gongracr) [reply logic/src/commonJvmAndroid/kotlin/com.wire.kalium.logic/util/BackupUtils.kt:1]: Let's leave these KMP functions on the same file for context reasons

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/ExtractCompressedFileUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    /**
+     * Extracts all the files from the provided compressed file path.
+     * @param compressedBackupFilePath The absolute file system path to the compressed file.
+     * @return A [ExtractCompressedBackupFileResult] indicating the Failure or Success with the root [Path] of the extracted files and a
+     * boolean indicating whether the returned paths belong to an encrypted or non-encrypted file.
+     */
+    suspend operator fun invoke(compressedBackupFilePath: Path): ExtractCompressedBackupFileResult
+}
+
+internal class ExtractCompressedFileUseCaseImpl(
```
[the remark this answers is not in the document]
Comment (gongracr) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/ExtractCompressedFileUseCase.kt]: Added them

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/CreateBackupUseCase.kt [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+interface CreateBackupUseCase {
+    /**
+     * Creates a compressed backup file, containing a metadata json file and the current state of the database. This file can be encrypted
+     * with the provided password if it is not empty. Otherwise, the file will be unencrypted.
+     * @param password The password to encrypt the backup file with. If empty, the file will be unencrypted.
+     */
+    suspend operator fun invoke(password: String): CreateBackupResult
+}
+
+internal class CreateBackupUseCaseImpl(
```
[the remark this answers is not in the document]
Comment (gongracr) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/CreateBackupUseCase.kt:46]: I would aim to merge this PR asap, as well as the AR one. And add these tests later on in order not to block this rocking feature for much longer.

**Review by ohassine — APPROVED** [review]

File: .github/workflows/gradle-android-tests.yml [hunk 7]
```diff
@@ -1,84 +1,86 @@
...
-        # Remove some files from the Gradle cache, so they aren't cached by GitHub Actions.
-        # Restoring these files from a GitHub Actions cache might cause problems for future builds.
-        run: |
-          rm -f ~/.gradle/caches/modules-2/modules-2.lock
-          rm -f ~/.gradle/caches/modules-2/gc.properties
+            - name: Cleanup Gradle Cache
+                # Remove some files from the Gradle cache, so they aren't cached by GitHub Actions.
+                # Restoring these files from a GitHub Actions cache might cause problems for future builds.
+              run: |
+                  rm -f ~/.gradle/caches/modules-2/modules-2.lock
```
Comment (vitorhugods) [inline .github/workflows/gradle-android-tests.yml:85]: Does the IDE like this 4 space indentation for YML files?

2 spaces is the standard for YML files and all of our other files in the project are 2 spaces as well.

Kinda annoying when moving code between files if the indentation doesn't match.

Comment (trOnk12) [reply .github/workflows/gradle-android-tests.yml:85]: @vitorhugods not sure what is happening I copy the content of this file from the develop changes it in 3 places and push it and this happens ;s

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/RestoreBackupUseCase.kt [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    private suspend fun runSanityChecks(extractedBackupPath: Path, password: String?): Either<Failure, Pair<Path?, Boolean>> =
+        if (password.isNullOrEmpty()) {
+            // Backup is not encrypted so we don't need to return the path to the encrypted file
+            checkIsValidAuthor(extractedBackupPath).fold({ Either.Left(it) }, { Either.Right(null to false) })
+        } else {
+            // If the backup is encrypted, the sanity checks are done when decoding the file
+            checkIsValidEncryption(extractedBackupPath)
+        }
+
+    private suspend fun checkIsValidEncryption(extractedBackupPath: Path): Either<Failure, Pair<Path?, Boolean>> = with(kaliumFileSystem) {
```
Comment (trOnk12) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/RestoreBackupUseCase.kt:116]: this is failing due to the extractedBackupPath pointing to the file, I have no idea what the intention was of checking if it contains the .cc20 file ? perhaps this is meant to be used within a zip file ?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/RestoreBackupUseCase.kt [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        }
+
+    private suspend fun decryptExtractAndImportBackup(
+        encryptedFilePath: Path?,
+        extractedBackupRootPath: Path,
+        password: String
+    ): RestoreBackupResult {
+        val backupSource = kaliumFileSystem.source(encryptedFilePath!!)
+        val extractedBackupPath = extractedBackupRootPath / BACKUP_UNENCRYPTED_FILE_NAME
+        val backupSink = kaliumFileSystem.sink(extractedBackupPath)
```
Comment (trOnk12) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/RestoreBackupUseCase.kt:74]: after the other place is fixed, this is failing due to extractedBackupPath pointing to a non-existing path

**Pull request merged**